services/flake_detection.py

```diff
@@ -1,67 +1,170 @@
 from collections import defaultdict
-from dataclasses import dataclass, field
+from dataclasses import dataclass
 from datetime import datetime, timedelta
-from enum import Enum
+from functools import cached_property
 from logging import getLogger
 from sys import getsizeof
-from typing import List, Set, Tuple
 
 from sentry_sdk import metrics, trace
 from test_results_parser import Outcome
 
+from database.enums import FlakeSymptomType
 from database.models.core import Commit, Repository
 from database.models.reports import CommitReport, Test, TestInstance, Upload
 
 log = getLogger(__name__)
 
+FlakeDetectionResult = dict[Test, set[FlakeSymptomType]]
 
-@dataclass
-class FlakeDetectionObject:
-    counter: int = 0
-    branch: Set[str] = field(default_factory=set)
-    time: List[datetime] = field(default_factory=list)
 
+class BaseSymptomDetector:
+    """
+    blueprint class for a single symptom of flakiness detector
 
-class FlakeType(Enum):
-    FAILED_IN_DEFAULT_BRANCH = "failed_in_default_branch"
-    CONSECUTIVE_DIFF_OUTCOMES = "consecutive_diff_outcomes"
-    UNRELATED_MATCHING_FAILURES = "unrelated_matching_failures"
+    classes that implement this interface are expected to:
```

Comment:
[very nit] I think this can be broken up into docstrings in the abstract methods directly... I feel that this could very easily become obsolete and people forget to update it.